src/compiler/utilities.ts

```diff
@@ -233,12 +262,12 @@ namespace ts {
                 case SyntaxKind.Block:
                     // function block is not considered block-scope container
                     // see comment in binder.ts: bind(...), case for SyntaxKind.Block
-                    if (!isFunctionLike(current.parent)) {
+                    if (!isFunctionLike(getAncestorOrSelf ? getAncestorOrSelf(offset + 1) : current.parent)) {
                         return current;
                     }
             }
 
-            current = current.parent;
+            current = getAncestorOrSelf ? getAncestorOrSelf(++offset) : current.parent;

```

Comment:
Perhaps I'm misunderstanding what `getAncestorOrSelf()` does, but let's suppose `offset === 25` or some large number, and `getAncestorOrSelf(offset)` finds not ancestor at that level. Should `getAncestorOrSelf()` be expected to return `node` or `current` in that case? How do we prevent infinite loops?